### PR TITLE
[clang] Inject tokens containing #embed back into token stream

### PR DESCRIPTION
**Pull request by Fznamznon**
Instead of playing "whack a mole" with places where #embed should be expanded as comma-separated list, just inject each byte as a token back into the stream, separated by commas.

### CODE REVIEW

File: clang/test/Preprocessor/embed_codegen.cpp [hunk 1]
```diff
@@ -43,8 +43,9 @@ a
 };
 
 // CHECK: store i32 107, ptr %b, align 4
-int b =
+int b = (
 #embed<jk.txt>
+    )
```
Comment (AaronBallman) [inline clang/test/Preprocessor/embed_codegen.cpp:48]: Oh nice, these changes fixed a bug -- this code should not have compiled before.

**Review by AaronBallman — APPROVED** [review]
In general, I like this! LGTM, though please give others a chance to review before landing.

File: clang/test/Preprocessor/embed_codegen.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/test/Preprocessor/embed_codegen.cpp:48]: In fact, they fix several bugs.

**Comment by Fznamznon** [comment]
It's been a week. Are we good to go? @cor3ntin

File: clang/lib/Parse/ParseExpr.cpp [hunk 2]
```diff
@@ -1018,6 +1018,7 @@ ExprResult Parser::ParseCastExpression(CastParseKind ParseKind,
 
     // primary-expression
   case tok::numeric_constant:
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseExpr.cpp:1020]: Can you adapt the couple other places where we parse a numeric constant?

Comment (Fznamznon) [reply clang/lib/Parse/ParseExpr.cpp:1020]: Could you please elaborate which other places you meant? I briefly examined all mentions of `tok::numeric_constant` and `ActOnNumericConstant` in `lib/Parse` and modifying any of these doesn't seem to be reasonable for `#embed` support.

**Comment by jakubjelinek** [comment]
I'd like to point out that most of the testcases I wrote for the GCC implementation still fail with the latest clang on godbolt.

**Comment by Fznamznon** [comment]
> I'd like to point out that most of the testcases I wrote for the GCC implementation still fail with the latest clang on godbolt.

This patch should help with at least some of them.

**Comment by jakubjelinek** [comment]
Maybe, but I've add quite a few new testcases too...

**Comment by AaronBallman** [comment]
> > I'd like to point out that most of the testcases I wrote for the GCC implementation still fail with the latest clang on godbolt.
> 
> This patch should help with at least some of them.

It'd be good to verify that it does help with them or see if there are other issues we're not yet tracking (but please do not copy the tests from the GCC implementation into Clang because of licensing).

**Comment by AaronBallman** [comment]
> I'd like to point out that most of the testcases I wrote for the GCC implementation still fail with the latest clang on godbolt.

Thank you for letting us know! The plan is to continue to polish this feature through the Clang 19 release. The branch date for us is in two weeks, so the more we can get done before the branch, the better.

File: clang/lib/Parse/ParseExpr.cpp (hunk 2 above)
Comment (Fznamznon) [reply clang/lib/Parse/ParseExpr.cpp:1020]: ping @cor3ntin

Comment (cor3ntin) [reply clang/lib/Parse/ParseExpr.cpp:1020]: These cases

```cpp
[[clang::availability(
   #embed "smth.txt"
)]] void f();

struct S {
   virtual void f() =
#embed "zero.bin"
    ;
};
```

However, it's a bit ambiguous whether these should be supported at all in the proposed C++ wording for embed. I sent a mail to the C++ committee

It's probably fine to ignore that for now, the patch is a great improvement as it is

File: clang/lib/Sema/SemaExpr.cpp [hunk 3]
```diff
@@ -3644,9 +3644,11 @@ bool Sema::CheckLoopHintExpr(Expr *E, SourceLocation Loc, bool AllowZero) {
 ExprResult Sema::ActOnNumericConstant(const Token &Tok, Scope *UDLScope) {
   // Fast path for a single digit (which is quite common).  A single digit
   // cannot have a trigraph, escaped newline, radix prefix, or suffix.
-  if (Tok.getLength() == 1) {
+  if (Tok.getLength() == 1 || Tok.getKind() == tok::binary_data) {
     const char Val = PP.getSpellingOfSingleCharacterNumericConstant(Tok);
-    return ActOnIntegerConstant(Tok.getLocation(), Val-'0');
+    return ActOnIntegerConstant(
+        Tok.getLocation(),
+        (Tok.getKind() == tok::binary_data) ? Val : Val - '0');
```
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaExpr.cpp]: Sure.

**Comment by Fznamznon** [comment]
I'm looking through https://gcc.gnu.org/pipermail/gcc-patches/2024-June/655012.html
This patch fixes cases like:

```
const unsigned char w[] = {
#embed __FILE__ prefix([0] = 42, [15] =) limit(32)
};
```

And also cases like
```
void foo (int, int, int, int);
void bar (void) { foo (
#embed __FILE__ limit (4) prefix (172 + ) suffix (+ 2)
); }
```
as intended, the test case similar to latter is modified by the patch.
There are still problems with `#embed ""`  `__if_empty__ ((({{[0[0{0{0(0(0)1)1}1}]]}}))))`, has_embed, handling of prefix/suffix trying to #embed "", https://godbolt.org/z/PsoTr5f7K  and etc. These I consider out of the scope of this patch and will post more details to https://github.com/llvm/llvm-project/issues/95222 .

**Comment by jakubjelinek** [comment]
I meant also e.g. basic tests like gcc/testsuite/c-c++-common/cpp/embed-20.c in
https://gcc.gnu.org/pipermail/gcc-patches/2024-July/657053.html
(and various others too).  Just checked that one and it still fails on latest clang trunk using godbolt.

**Comment by jakubjelinek** [comment]
@ThePhd @AaronBallman
And even more importantly (checking on godbolt again):
```c
int a = sizeof (
    #embed __FILE__ limit (1)
);
```
is 1 in clang as well as clang++ trunk and 4 with clang/clang++ trunk with -save-temps.
I thought there was agreement that at least for C the literals have type int, for C++ it is fuzzy and depends on what will be voted in but in any case, there shouldn't be different code produced between integrated preprocessing and -save-temps.
If C++ requires some cast, it would need to make it clear what that exact cast is, i.e. whether it is supposed to be
```c
12,143,12,16
```
or
```c
static_cast<unsigned char>(12),static_cast<unsigned char>(143),static_cast<unsigned char>(12),static_cast<unsigned char>(16)
```
or
```c
(unsigned char)12,(unsigned char)143,(unsigned char)12,(unsigned char)16
```
or
```c
'\014','\0217','\014','\020'
```
or whatever else and the preprocessor would need to emit that cast on every literal (unless using some extension like #embed "." __gnu__::__base64__("...") that the GCC patchset uses for the inner parts of the longer sequences.
I think the exact type of cast can affect parsing of some of the expressions, e.g.
```c
extern long long p[];
auto foo () {
return
#embed __FILE__ limit (1)
[p];
}
```
will behave one way with the cast is (unsigned char)12 and differently if it is static_cast<unsigned char>(12) or ((unsigned char)12).
Most of the other bugs I'm seeing are also about consistency, with -save-temps it works IMHO correctly while without it misbehaves.  The behavior has to be the same.

**Comment by Fznamznon** [comment]
> Just checked that one and it still fails on latest clang trunk using godbolt.

Hmm, ok. https://github.com/llvm/llvm-project/pull/99023 . Thanks for pointing this out.

**Comment by AaronBallman** [comment]
> I thought there was agreement that at least for C the literals have type int,

That's correct; this patch gets started on addressing that issue by updating the way we expand the embedded contents but isn't intended to fix the type issue yet (that will be a follow-up patch). When we do the follow-up, we should ensure the embedded elements are of type `int` (for both C and C++; we don't want different types coming out of the preprocessor depending on language mode as that will make life harder for mixed C and C++ code bases).

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by ThePhD** [comment]
Just to be clear, I'm editing the C++ wording to also produce integer constant expressions that should be `int` typed, so both languages should be in sync.

**Comment by Fznamznon** [comment]
FYI https://github.com/llvm/llvm-project/pull/99050

File: clang/include/clang/Lex/Preprocessor.h [hunk 4]
```diff
@@ -2123,17 +2123,18 @@ class Preprocessor {
   char
   getSpellingOfSingleCharacterNumericConstant(const Token &Tok,
```
Comment (bnbarham) [inline clang/include/clang/Lex/Preprocessor.h:2124]: Both the documentation and name of this function is incorrect now that the underlying functionality has changed - would you mind updating it to something more appropriate @Fznamznon? It used to return the character itself:
```
  /// Given a Token \p Tok that is a numeric constant with length 1,
  /// return the character.
```

But now returns the value of the character.

Comment (AaronBallman) [reply clang/include/clang/Lex/Preprocessor.h:2124]: I've updated the comment and the interface in 295e4f49aefb2b07501be9f845df598b3ee280f1

Comment (Fznamznon) [reply clang/include/clang/Lex/Preprocessor.h:2124]: Thank you @AaronBallman

Comment (bnbarham) [reply clang/include/clang/Lex/Preprocessor.h:2124]: Thanks @AaronBallman!

**Pull request merged**